src/pybind/mgr/dashboard/frontend/src/app/shared/directives/auth-storage.directive.ts

```diff
@@ -0,0 +1,93 @@
+import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import _ from 'lodash';
+
+import { Permission, Permissions } from '~/app/shared/models/permissions';
+import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
+
+@Directive({
+  selector: '[cdAuthStorage]'
+})
+export class AuthStorageDirective {
+  permissions: Permissions;
+
+  constructor(
+    private templateRef: TemplateRef<any>,
+    private viewContainer: ViewContainerRef,
+    private authStorageService: AuthStorageService
+  ) {}
+
+  @Input('cdAuthStorage') set cdAuthStorage(condition: string | string[] | object) {
+    this.permissions = this.authStorageService.getPermissions();
+    if (this.isAuthorized(condition)) {
+      this.viewContainer.createEmbeddedView(this.templateRef);
+    } else {
+      this.viewContainer.clear();
+    }
+  }
+
+  private isAuthorized(condition: any): boolean {
+    /*
+      single permission equivalent to check e.g osd.
+      We check if the permission key exists on the Permissions class then
+      if true We check for read permission and display the component accordingly
+    */
+
+    if (_.isString(condition)) {
+      if (_.has(this.permissions, [condition, 'read']) && this.permissions[condition]['read']) {
+        return true;
+      }
+    }
+
+    /*
+      lodash _isArray matches both objects and arrays.
+      By checking if the first index contains a value we can tell if it is truly an array or not
+      Also works for _isObject
+    */
+
+    if (_.isArray(condition) && condition[0] !== undefined) {
+      const checkedPermissions: boolean[] = [];
+
+      _.forEach(condition, (value) => {
+        if (_.has(this.permissions, [value, 'read']) && this.permissions[value]['read']) {
+          checkedPermissions.push(true);
+        } else {
+          checkedPermissions.push(false);
+        }
+      });
+
+      /*
+        check if all permissions in the checkedPermissions array are true.
+        if no false values are found then requested permissions are granted so we display the component
+      */
+      if (!checkedPermissions.includes(false)) {
+        return true;
+      }
+    }
+
+    if (_.isObject(condition) && condition[0] === undefined) {
+      const checkedPermissions: boolean[] = [];
+
+      _.forEach(condition, (value: Permission | Permission[], key: keyof Permissions) => {
+        _.forEach(value, (val) => {
+          if (_.has(this.permissions, [key, val]) && this.permissions[key][val]) {
+            checkedPermissions.push(true);
+          } else {
+            checkedPermissions.push(false);
+          }
+        });
+      });
+
+      /*
+        check if all permissions in the checkedPermissions array are true.
+        if no false values are found then requested permissions are granted so we display the component
+      */
+
+      if (!checkedPermissions.includes(false)) {
+        return true;
+      }
+    }
```

Comment:
And same here, but in this case you'll have to figure out how to deal with nested arrays! :D

Reply:
_every works here too. In this case, I had to nest two of them to evaluate the internal arrays in the object permission

Reply:
Looks really good now! When you implement the `matchAll`, you'll have to dynamically switch from `_.some` to `_.every` in order to provide the 2 modes (loose/matchAny and strict/matchAll).